File: Src/java/engine-fhir/src/main/java/org/opencds/cqf/cql/engine/fhir/retrieve/R4FhirQueryGenerator.java
package org.opencds.cqf.cql.engine.fhir.retrieve;

import java.net.URLDecoder;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import ca.uhn.fhir.context.FhirVersionEnum;
import org.hl7.fhir.instance.model.api.IBaseConformance;
import org.hl7.fhir.instance.model.api.ICompositeType;
import org.hl7.fhir.r4.model.CapabilityStatement;
import org.hl7.fhir.r4.model.Coding;
import org.hl7.fhir.r4.model.DataRequirement;
import org.hl7.fhir.r4.model.DateTimeType;
import org.hl7.fhir.r4.model.Duration;
import org.hl7.fhir.r4.model.Period;
import org.hl7.fhir.r4.model.Type;
import org.opencds.cqf.cql.engine.elm.execution.SubtractEvaluator;
import org.opencds.cqf.cql.engine.fhir.exception.FhirVersionMisMatchException;
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterMap;
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver;
import org.opencds.cqf.cql.engine.model.ModelResolver;
import org.opencds.cqf.cql.engine.runtime.Code;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Interval;
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;


public class R4FhirQueryGenerator extends BaseFhirQueryGenerator {
    public R4FhirQueryGenerator(SearchParameterResolver searchParameterResolver, TerminologyProvider terminologyProvider,
                                ModelResolver modelResolver) throws FhirVersionMisMatchException {
        super(searchParameterResolver, terminologyProvider, modelResolver, searchParameterResolver.getFhirContext());
    }

    @Override
    public FhirVersionEnum getFhirVersion() {
        return FhirVersionEnum.R4;
    }

    @Override
    public List<String> generateFhirQueries(ICompositeType dreq, DateTime evaluationDateTime, Map<String, Object> contextValues, Map<String, Object> parameters, IBaseConformance capStatement) {
        if (!(dreq instanceof DataRequirement)) {
            throw new IllegalArgumentException("dataRequirement argument must be a DataRequirement");
        }
        if (capStatement != null && !(capStatement instanceof CapabilityStatement)) {
            throw new IllegalArgumentException("capabilityStatement argument must be a CapabilityStatement");
        }

        DataRequirement dataRequirement = (DataRequirement)dreq;

        List<String> queries = new ArrayList<>();

        List<CodeFilter> codeFilters = new ArrayList<CodeFilter>();
        if (dataRequirement.hasCodeFilter()) {
            for (DataRequirement.DataRequirementCodeFilterComponent codeFilterComponent : dataRequirement.getCodeFilter()) {
                if (!codeFilterComponent.hasPath()) continue;

                String codePath = null;
                List<Code> codes = null;
                String valueSet = null;

                codePath = codeFilterComponent.getPath();

                if (codeFilterComponent.hasValueSetElement()) {
                    valueSet = codeFilterComponent.getValueSet();
                }

                if (codeFilterComponent.hasCode()) {
                    codes = new ArrayList<Code>();

                    List<Coding> codeFilterValueCodings = codeFilterComponent.getCode();
                    for (Coding coding : codeFilterValueCodings) {
                        if (coding.hasCode()) {
                            Code code = new Code();
                            code.setSystem(coding.getSystem());
                            code.setCode(coding.getCode());
                            codes.add(code);
                        }
                    }
                }

                codeFilters.add(new CodeFilter(codePath, codes, valueSet));
            }
        }

        List<DateFilter> dateFilters = new ArrayList<DateFilter>();

        if (dataRequirement.hasDateFilter()) {
            for (DataRequirement.DataRequirementDateFilterComponent dateFilterComponent : dataRequirement.getDateFilter()) {
                String datePath = null;
                String dateLowPath = null;
                String dateHighPath = null;
                Interval dateRange = null;

                if (dateFilterComponent.hasPath() && dateFilterComponent.hasSearchParam()) {
                    throw new UnsupportedOperationException("Either a path or a searchParam must be provided, but not both");
                }

                if (dateFilterComponent.hasPath()) {
                    datePath = dateFilterComponent.getPath();
                } else if (dateFilterComponent.hasSearchParam()) {
                    datePath = dateFilterComponent.getSearchParam();
                }

                // TODO: Deal with the case that the value is expressed as an expression extension
                if (dateFilterComponent.hasValue()) {
                    Type dateFilterValue = dateFilterComponent.getValue();
                    if (dateFilterValue instanceof DateTimeType && dateFilterValue.hasPrimitiveValue()) {
                        dateLowPath = "valueDateTime";
                        dateHighPath = "valueDateTime";
                        String offsetDateTimeString = ((DateTimeType)dateFilterValue).getValueAsString();
                        DateTime dateTime = new DateTime(OffsetDateTime.parse(offsetDateTimeString));

                        dateRange = new Interval(dateTime, true, dateTime, true);
                    } else if (dateFilterValue instanceof Duration && ((Duration)dateFilterValue).hasValue()) {
                        // If a Duration is specified, the filter will return only those data items that fall within Duration before now.
                        Duration dateFilterAsDuration = (Duration)dateFilterValue;

                        org.opencds.cqf.cql.engine.runtime.Quantity dateFilterDurationAsCQLQuantity =
                            new org.opencds.cqf.cql.engine.runtime.Quantity().withValue(dateFilterAsDuration.getValue()).withUnit(dateFilterAsDuration.getUnit());

                        DateTime diff = ((DateTime)SubtractEvaluator.subtract(evaluationDateTime, dateFilterDurationAsCQLQuantity));

                        dateRange = new Interval(diff, true, evaluationDateTime, true);
                    } else if (dateFilterValue instanceof Period && ((Period)dateFilterValue).hasStart() && ((Period)dateFilterValue).hasEnd()) {
                        dateLowPath = "valueDateTime";
                        dateHighPath = "valueDateTime";
                        dateRange = new Interval(((Period)dateFilterValue).getStart(), true, ((Period)dateFilterValue).getEnd(), true);
                    }

                    dateFilters.add(new DateFilter(datePath, dateLowPath, dateHighPath, dateRange));
                }
            }
        }

        // Patient is the default context if one is not specified
        String contextType = "Patient";

        if (dataRequirement.hasSubjectReference()) {
            throw new IllegalArgumentException("Cannot process data requirements with subjects specified as references");
        }

        if (dataRequirement.hasSubjectCodeableConcept()) {
            Coding c = null;
            for (Coding coding : dataRequirement.getSubjectCodeableConcept().getCoding()) {
                if ("http://hl7.org/fhir/resource-types".equals(coding.getSystem())) {
                    c = coding;
                    break;
                }
            }
            if (c == null) {
                throw new IllegalArgumentException("Cannot process data requirements with subjects not specified using a code from http://hl7.org/fhir/resource-types");
            }
            contextType = c.getCode();
        }

        Object contextPath = modelResolver.getContextPath(contextType, dataRequirement.getType());
        Object contextValue = contextValues.get(contextType);
        String templateId = dataRequirement.getProfile() != null && !dataRequirement.getProfile().isEmpty()
            ? dataRequirement.getProfile().get(0).getValue()
            : null;

            List<SearchParameterMap> maps = setupQueries(contextType, (String)contextPath, contextValue, dataRequirement.getType(), templateId,
            codeFilters, dateFilters);

        for (SearchParameterMap map : maps) {
            String query = null;
            try {
                query = URLDecoder.decode(map.toNormalizedQueryString(fhirContext), "UTF-8");
            } catch (Exception ex) {
                query = map.toNormalizedQueryString(fhirContext);
            }
            queries.add(dataRequirement.getType() + query);
        }

        return queries;
    }
}

File: Src/java/engine-fhir/src/main/kotlin/org/opencds/cqf/cql/engine/fhir/retrieve/R4FhirQueryGenerator.kt
package org.opencds.cqf.cql.engine.fhir.retrieve

import ca.uhn.fhir.context.FhirVersionEnum
import java.net.URLDecoder
import java.time.OffsetDateTime
import org.hl7.fhir.instance.model.api.IBaseConformance
import org.hl7.fhir.instance.model.api.ICompositeType
import org.hl7.fhir.r4.model.CapabilityStatement
import org.hl7.fhir.r4.model.Coding
import org.hl7.fhir.r4.model.DataRequirement
import org.hl7.fhir.r4.model.DateTimeType
import org.hl7.fhir.r4.model.Duration
import org.hl7.fhir.r4.model.Period
import org.opencds.cqf.cql.engine.elm.executing.SubtractEvaluator
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver
import org.opencds.cqf.cql.engine.model.ModelResolver
import org.opencds.cqf.cql.engine.runtime.Code
import org.opencds.cqf.cql.engine.runtime.DateTime
import org.opencds.cqf.cql.engine.runtime.Interval
import org.opencds.cqf.cql.engine.runtime.Quantity
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider

class R4FhirQueryGenerator(
    searchParameterResolver: SearchParameterResolver,
    terminologyProvider: TerminologyProvider?,
    modelResolver: ModelResolver,
) :
    BaseFhirQueryGenerator(
        searchParameterResolver,
        terminologyProvider,
        modelResolver,
        searchParameterResolver.fhirContext,
    ) {

    override val fhirVersion = FhirVersionEnum.R4

    override fun generateFhirQueries(
        dataRequirement: ICompositeType?,
        evaluationDateTime: DateTime?,
        contextValues: MutableMap<String, Any?>?,
        parameters: MutableMap<String, Any?>?,
        capabilityStatement: IBaseConformance?,
    ): MutableList<String> {
        require(dataRequirement is DataRequirement) {
            "dataRequirement argument must be a DataRequirement"
        }
        require(!(capabilityStatement != null && capabilityStatement !is CapabilityStatement)) {
            "capabilityStatement argument must be a CapabilityStatement"
        }

        val queries: MutableList<String> = ArrayList()

        val codeFilters: MutableList<CodeFilter> = ArrayList()
        if (dataRequirement.hasCodeFilter()) {
            for (codeFilterComponent in dataRequirement.getCodeFilter()) {
                if (!codeFilterComponent.hasPath()) continue
                var codes: MutableList<Code>? = null
                var valueSet: String? = null

                val codePath = codeFilterComponent.getPath()

                if (codeFilterComponent.hasValueSetElement()) {
                    valueSet = codeFilterComponent.getValueSet()
                }

                if (codeFilterComponent.hasCode()) {
                    codes = ArrayList()

                    val codeFilterValueCodings = codeFilterComponent.getCode()
                    for (coding in codeFilterValueCodings) {
                        if (coding.hasCode()) {
                            val code = Code()
                            code.system = coding.getSystem()
                            code.code = coding.getCode()
                            codes.add(code)
                        }
                    }
                }

                codeFilters.add(CodeFilter(codePath, codes, valueSet))
            }
        }

        val dateFilters: MutableList<DateFilter> = ArrayList()

        if (dataRequirement.hasDateFilter()) {
            for (dateFilterComponent in dataRequirement.getDateFilter()) {
                var datePath: String? = null
                var dateLowPath: String? = null
                var dateHighPath: String? = null
                var dateRange: Interval? = null

                if (dateFilterComponent.hasPath() && dateFilterComponent.hasSearchParam()) {
                    throw UnsupportedOperationException(
                        "Either a path or a searchParam must be provided, but not both"
                    )
                }

                if (dateFilterComponent.hasPath()) {
                    datePath = dateFilterComponent.getPath()
                } else if (dateFilterComponent.hasSearchParam()) {
                    datePath = dateFilterComponent.getSearchParam()
                }

                // TODO: Deal with the case that the value is expressed as an expression extension
                if (dateFilterComponent.hasValue()) {
                    when (val dateFilterValue = dateFilterComponent.getValue()) {
                        is DateTimeType if dateFilterValue.hasPrimitiveValue() -> {
                            dateLowPath = "valueDateTime"
                            dateHighPath = "valueDateTime"
                            val offsetDateTimeString = dateFilterValue.valueAsString
                            val dateTime = DateTime(OffsetDateTime.parse(offsetDateTimeString))

                            dateRange = Interval(dateTime, true, dateTime, true)
                        }

                        is Duration if dateFilterValue.hasValue() -> {
                            // If a Duration is specified, the filter will return only those data
                            // items that fall within
                            // Duration before now.
                            val dateFilterAsDuration = dateFilterValue

                            val dateFilterDurationAsCQLQuantity =
                                Quantity()
                                    .withValue(dateFilterAsDuration.getValue())
                                    .withUnit(dateFilterAsDuration.getUnit())

                            // Passing null as the state argument to the subtract method is fine
                            // here since that method only uses the state when it has to convert
                            // Quantities with different units which cannot happen here.
                            val diff =
                                (SubtractEvaluator.subtract(
                                    evaluationDateTime,
                                    dateFilterDurationAsCQLQuantity,
                                    null,
                                ) as DateTime?)

                            dateRange = Interval(diff, true, evaluationDateTime, true)
                        }

                        is Period if dateFilterValue.hasStart() && dateFilterValue.hasEnd() -> {
                            dateLowPath = "valueDateTime"
                            dateHighPath = "valueDateTime"
                            dateRange =
                                Interval(
                                    dateFilterValue.getStart(),
                                    true,
                                    dateFilterValue.getEnd(),
                                    true,
                                )
                        }
                    }

                    dateFilters.add(DateFilter(datePath, dateLowPath, dateHighPath, dateRange))
                }
            }
        }

        // Patient is the default context if one is not specified
        var contextType: String? = "Patient"

        require(!dataRequirement.hasSubjectReference()) {
            "Cannot process data requirements with subjects specified as references"
        }

        if (dataRequirement.hasSubjectCodeableConcept()) {
            var c: Coding? = null
            for (coding in dataRequirement.subjectCodeableConcept.getCoding()) {
                if ("http://hl7.org/fhir/resource-types" == coding.getSystem()) {
                    c = coding
                    break
                }
            }
            requireNotNull(c) {
                "Cannot process data requirements with subjects not specified using a code from http://hl7.org/fhir/resource-types"
            }
            contextType = c.getCode()
        }

        val contextPath = modelResolver.getContextPath(contextType, dataRequirement.getType())
        val contextValue = contextValues?.get(contextType)
        val templateId =
            if (dataRequirement.getProfile() != null && !dataRequirement.getProfile().isEmpty())
                dataRequirement.getProfile()[0].value
            else null

        val maps =
            setupQueries(
                contextType,
                contextPath as String?,
                contextValue,
                dataRequirement.getType(),
                templateId,
                codeFilters,
                dateFilters,
            )

        for (map in maps) {
            val query =
                try {
                    URLDecoder.decode(map.toNormalizedQueryString(fhirContext), "UTF-8")
                } catch (ex: Exception) {
                    map.toNormalizedQueryString(fhirContext)
                }
            queries.add(dataRequirement.getType() + query)
        }

        return queries
    }
}
